### PR TITLE
Add unittest framework

### PR DESCRIPTION
**Pull request by stin7**
I created a new branch for unit test suite, per issue #21. New code resides in folder outside of `backend`, so this branch should be safe to bring into master.

I decided to submit this pull request before all tests are written (most are not) so that others can help build them out.

### CODE REVIEW

**Comment by stin7** [comment]
Quick follow up - I misspoke, there was one change to `backend`: I added a blank `__init__.py` file.

Also, I thought my pull request would be for a new branch called `add-unittest-framework` but it looks like this is set to go straight into `master`. Let me know if I should redo anything on my end, I'm fairly new to Forking and Pull Requests on Github.

Cheers,
Austin

**Comment by brad** [comment]
It looks like you did it right, although it's unfortunately not merge-able right now. I'm guessing it's a conflict in README.md. If you get a chance, try and merge MayOneUS/master into your branch and see if you can resolve the conflicts. I'll take a look at this as soon as I get a chance.

**Comment by stin7** [comment]
It was README. I pulled the current master into my branch and resolved the conflict (dropped my changes)

If this pull request uses the current state of the branch it should be good to merge. If not I can resubmit a new pull request.

Cheers,
Austin

File: testrunner.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/python
+import sys
+# Install the Python unittest2 package before you run this script.
```
Comment (brad) [inline testrunner.py]: The `unittest` module is built into python, isn't it? I believe this comment is misleading.

File: testrunner.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  sys.path.insert(0, sdk_path)
+  sys.path.insert(1,'backend')
+  sys.path.insert(1,'backend/stripe')
+  import dev_appserver
+  dev_appserver.fix_sys_path()
+  suite = unittest.loader.TestLoader().discover(test_path)
+  unittest.TextTestRunner(verbosity=2).run(suite)
+
+if __name__ == '__main__':
+  print('UNITTESTS FOR MAYONE APPENGINE')
```
Comment (brad) [inline testrunner.py]: I don't think this or the output on line 29 are necessary, the tests can speak for themselves.

File: testrunner.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  SDK_PATH = SDK_PATH_manual
+  TEST_PATH = TEST_PATH_manual
+  SDK_PATH_override = raw_input('Enter SDK_PATH Override (Leave blank to use the default): ')
```
Comment (brad) [inline testrunner.py]: I would much prefer if this variable was obtained from `sys.argv`. That way the tests can be ran unattended.

File: unittests/test_Main.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+#from datetime import datetime, timedelta
+
+#from google.appengine.api import memcache
+from google.appengine.ext import ndb
+from google.appengine.ext import testbed
+
+from backend import main
+
+class TestGetTotalHandler(unittest.TestCase):
+    def setUp(self):
```
Comment (brad) [inline unittests/test_Main.py]: Please stick with a two space indent consistently throughout this file.

File: unittests/test_Main.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      # Then activate the testbed, which prepares the service stubs for use.
+      self.testbed.activate()
+      # Next, declare which service stubs you want to use.
+      self.testbed.init_datastore_v3_stub()
+
+    def tearDown(self):
+      self.testbed.deactivate()
+     
+    def test_get(self):
+      self.fail('TEST NOT IMPLEMENTED')
```
Comment (brad) [inline unittests/test_Main.py]: I don't feel great about committing a bunch of failing tests. Can we leave them out until they do something/pass?

File: unittests/test_Model.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    logging.info('Test updating that user we just created...')
+    user1 = model.User.createOrUpdate(email=fake_email, occupation='Regulator')
+    self.assertEqual(user1.occupation, 'Regulator')
+    
+    #TODO: confirm storage in datastore.
+    #TODO: see if we can store bad emails and other data
+
+
+class TestPledge(unittest.TestCase):
+    def setUp(self):
```
Comment (brad) [inline unittests/test_Model.py]: Again, please stick with a two space indent throughout this file.

File: unittests/test_Main.py (hunk 5 above)
Comment (erinzm) [inline unittests/test_Main.py]: No, do commit failing tests. They are a big part of testing; as a template, they help you write code to _make the tests pass_.

Comment (brad) [inline unittests/test_Main.py]: OK, sure, I'm with you. It's good to have that encouragement to get them done.

Comment (hjfreyer) [inline unittests/test_Main.py]: FWIW, I disagree. I do TDD, but you don't break the branch for others.

**Comment by brad** [comment]
@stin7 I have made some inline comments requesting changes. But this looks really great. I ran the tests locally, everything ran smoothly.

**Comment by stin7** [comment]
Great - I've fixed the indentation, removed extraneous comments and print outs, and am pulling the first passed argument (if it exists) as an override for the SDK_PATH

**Comment by brad** [comment]
You know, Hunter is right. I don't know if you saw his inline comment, but let's just start with the working tests. Failing tests are good on your local machine, but they shouldn't affect the rest of us. For the why, imagine the following scenario: I am working on feature X, so I write tests to go along with it. If there are still failing tests after I get my tests working, I might think my feature had an inadvertent effect on the code base and go searching for a way to fix it.
No, better to just keep the failing tests to yourself and commit them when they pass.

**Comment by stin7** [comment]
Sounds good to me - will remove the fail assertions. Should they be replaced with a logged warning to remind us of incomplete coverage? Or just leave it to us to remember and get to full coverage by some other means? I'm fine with either.

**Comment by brad** [comment]
Just remove them completely for now. @hjfreyer @jtolds  I've got some experience with Travis ci and coveralls.io. What do you think about integrating with those services to have some continuous integration with coverage reports?

**Comment by erinzm** [comment]
Sounds good; does anyone want Travis/Coveralls integration with IRC #mayone?

**Comment by jtolio** [comment]
that sounds awesome @brad and @ArchimedesPi. love me those build|passing tags :)

**Comment by erinzm** [comment]
Let's do Travis; write a make/rakefile that runs the tests, pushes to Coveralls, grabs the local coveralls results, and connects to `#mayone` and posts up the tests as `@travis`. Sound good?
(edit: there's a Coveralls/Travis framework for python tests here: https://github.com/coagulant/coveralls-python)

**Comment by brad** [comment]
Yeah except it can be done in .travis.yml so there's no need for a
[mr]akefile. I'm thinking it is a separate issue from this one. Let's get
this taken care of first.

**Comment by stin7** [comment]
LMK if there's anything else I can do on this

**Comment by brad** [comment]
Thank you @stin7 !
Looks great

**Pull request merged**